feat(ossm): adds OSSM annotations to the relevant cluster resources

cam-garrison opened the pull request:
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
Closes: #1087 

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

This change takes care of populating the following annotations to resources created through ODH Dashboard

- `opendatahub.io/service-mesh` can be set to `true` or `false` and will be used to adjust ODH components and make them part of the Service Mesh. This is applied on both `Project` and `Notebook` resources
- The dashboard will now link to notebooks through the mesh if feature flag `disableServiceMesh=false`

In addition, the code hides service-mesh-specific changes to annotations behind `disableServiceMesh` flag which is added in https://github.com/opendatahub-io/opendatahub-operator/pull/217

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

We have tested this manually to ensure that when `disableServiceMesh=false` the resources mentioned above contain the corresponding annotations. Also have tested manually to ensure that when `disableServiceMesh=false` notebook links send the user through the mesh ingress-gateway, and when `disableServiceMesh=true` the user is sent through the OpenShift route as before. 

The cluster adjustments to enable Service Mesh with ODH are still in progress (PR in manifests [here](https://github.com/opendatahub-io/odh-manifests/pull/818)) .

One visible change even without the Service Mesh deployed is that when `disableServiceMesh=false` the oauth-proxy sidecar is not injected anymore. This is intended.

Until this [PR](https://github.com/opendatahub-io/opendatahub-operator/pull/217) is merged and a new bundle is released you can use my build of the ODH Operator that includes the `disableServiceMesh` flag as part of the `ODHDashboardConfig`. Do so by running `operator-sdk run bundle quay.io/cgarriso/opendatahub-operator-bundle:dev-0.0.2 --namespace $OPERATOR_NAMESPACE` (req operator-sdk v1.24.1 )

Either build and push this dashboard image or use `quay.io/maistra-dev/odh-dashboard:ossm_annotations` image.

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->

If you can offer us some hints on how to test feature flags and resource creation through Jest tests we are happy to extend this PR with it.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits have meaningful messages (squashes happen on merge by the bot).
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

andrewballantyne commented:
/ok-to-test

cam-garrison commented:
I also have this [recorded demo](https://drive.google.com/file/d/1eMmxb2f7kp4GBkaQohvJ9ZMjFgsZ1VtU/view?usp=sharing) showing the changes which may be useful for review.

lucferbux commented on backend/src/routes/api/namespaces/namespaceUtils.ts:
```diff
@@ -60,11 +61,14 @@ export const applyNamespaceChange = async (
     case NamespaceApplicationCase.DSG_CREATION:
       labels = {
         'opendatahub.io/dashboard': 'true',
+        'opendatahub.io/service-mesh': String(!disableServiceMesh),
```
There's an underlying issue with this logic, what happens with projects that are created with service-mesh disabled that are already running?
I saw the demo and the logic on the code, but I'm still unsure about hat specific behaviour.

bartoszmajsak replied:
This is a very valid point @lucferbux, thanks for bringing this up. I think this deserves a deeper conversation as we would like to understand possible use cases here. Who is/will be in control of enabling service mesh? Is it the Dashboard? 

We might be missing some more orchestration logic, so we should figure out what are the paths here and come up with improvements.

But I think it has a broader scope, for example Kubeflow has env variable `USE_ISTIO=true` which will result in controllers creating `VirtualService`s. 

/cc @cfchase

lucferbux commented on backend/src/routes/api/notebooks/utils.ts:
```diff
@@ -27,10 +29,19 @@ export const getNotebookStatus = async (
-    });
+    const disableServiceMesh = getDashboardConfig().spec.dashboardConfig.disableServiceMesh;
+    let route: Route;
+    if (!disableServiceMesh) {
```
Here I'm seeing that same problem, this is based on the current status of the `OdhDashboardConfig`, but what happens when you enable/disable this feature and you've already deployed notebooks and projects?

lucferbux replied:
Because based on this piece of code, just based on the dashboard config, it will either try to get one route or the other, but this might happen for every notebook, not just the ones deployed with the feature enabled, would the itsio controller create routes for every notebook, even the ones. without the labels?

cam-garrison replied:
Copied it to the backend and added it here 93aa3ae

lucferbux commented on frontend/src/pages/projects/notebook/useRouteForNotebook.ts:
```diff
@@ -19,12 +21,17 @@ const useRouteForNotebook = (
       if (notebookName && projectName) {
-        getRoute(notebookName, projectName)
+        // if not using service mesh fetch openshift route, otherwise get Istio Ingress Gateway route
+        const getRoutePromise = dashboardConfig.spec.dashboardConfig.disableServiceMesh
```
I have the same exact question as the logic in the backend, we are mixing a solution of annotations and logic, is this covering all these scenarios:

* Enabling service mesh and trying to get the route of old notebooks without annotation
* Deploy new notebooks under service mesh, disabling the feature and getting the route

bartoszmajsak replied:
Good point, I think we shouldn't be checking the flag only. We should also check if the ns have the mesh annotation set. If both are true we should change the route to istio ingress. Otherwise assume old logic.

On that note, however: this will only work for users who have access to query `istio-system` for it. We are looking at a better way. Perhaps setting `'opendatahub.io/link'` annotation on the namespace by our controller should be sufficient. WDYT @lucferbux?

lucferbux replied:
Perfect, I saw recent updates, let me know when the pr is ready to review again and I'll look at it, thank you very much.

lucferbux reviewed (changes requested):
I'm requesting changes, this could be solved just with some clarification about the code, I'm lacking of the context about how it will be handled by the controller, but so far I'm worried about what happens when you enable/disable the feature flag with existing notebooks, which is the most common scenario we'll get.

lucferbux commented:
@bartoszmajsak @cam-garrison Hi! We've added new features that might conflict with the current PR, if you have any questions let me know and we can get through them.

bartoszmajsak commented on frontend/src/k8sTypes.ts:
```diff
@@ -286,6 +287,15 @@ export type ProjectKind = K8sResourceCommon & {
+  metadata: {
+    annotations?: DisplayNameAnnotations;
+  };
+  status?: {
+    phase: 'Active' | 'Terminating';
+  };
```
@cam-garrison is this something we need?

cam-garrison replied:
Added for use when fetching namespace annotation [here](https://github.com/opendatahub-io/odh-dashboard/pull/1088/files/dd22cd76e6aa2e2158829365ca209949c0549264#diff-05d6c60cadfff01f00f888ddbd48604ebd587d0a37f5678b713d653e19e4b973R22)

lucferbux replied:
No, just use `ProjectKind` instead, you get pretty much the same and openshift permissions are better to handle with projects

cam-garrison replied:
+1 https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/8376c2818e44bd39b6f995b25aaa4df9bdacea5a 

removed in this commit for using ProjectKind

cam-garrison commented:
Hi @lucferbux,

This PR is now ready for your re-review.

Changes since you last reviewed:

- Added a [service-mesh annotation](https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/1408c29db665cd881e2ef4c65d1a7208af9c70f9) to the created data science projects. This annotation is utilized in our [odh-project-controller](https://github.com/maistra/odh-project-controller) to add service-mesh resources and annotations to data science projects).
- Modified the [notebook linking logic](https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/47e43acb17cc8bc5541847a8da7a42c9e128c83a). Now, instead of fetching the ingress-gateway route from the `istio-system` namespace, we are fetching the host from an annotation created by the odh-project-controller. This change simplifies the logic when linking to notebooks and prevents permissions issues when trying to fetch a route from the `istio-system` namespace.
- Regarding migration - notebooks that were initially created with `disableServiceMesh=true` and later had `disableServiceMesh` set to false will have their annotations (inject-oauth and service-mesh) "migrated" upon restart. The commit adding this change can be found [here](https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/dd22cd76e6aa2e2158829365ca209949c0549264). Please note, this PR does not touch the data science project annotation; this task will be handled elsewhere (currently in our init-job in manifests we will switch all previously created dsp to be a part of the service mesh).

I believe these modifications, particularly concerning the migration of old notebooks' annotations, should address your prior concerns. Please take a look and don't hesitate to reach out if you have any questions!

lucferbux commented:
> Hi @lucferbux,
> 
> This PR is now ready for your re-review.
> 
> Changes since you last reviewed:
> 
> * Added a [service-mesh annotation](https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/1408c29db665cd881e2ef4c65d1a7208af9c70f9) to the created data science projects. This annotation is utilized in our [odh-project-controller](https://github.com/maistra/odh-project-controller) to add service-mesh resources and annotations to data science projects).
> * Modified the [notebook linking logic](https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/47e43acb17cc8bc5541847a8da7a42c9e128c83a). Now, instead of fetching the ingress-gateway route from the `istio-system` namespace, we are fetching the host from an annotation created by the odh-project-controller. This change simplifies the logic when linking to notebooks and prevents permissions issues when trying to fetch a route from the `istio-system` namespace.
> * Regarding migration - notebooks that were initially created with `disableServiceMesh=true` and later had `disableServiceMesh` set to false will have their annotations (inject-oauth and service-mesh) "migrated" upon restart. The commit adding this change can be found [here](https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/dd22cd76e6aa2e2158829365ca209949c0549264). Please note, this PR does not touch the data science project annotation; this task will be handled elsewhere (currently in our init-job in manifests we will switch all previously created dsp to be a part of the service mesh).
> 
> I believe these modifications, particularly concerning the migration of old notebooks' annotations, should address your prior concerns. Please take a look and don't hesitate to reach out if you have any questions!

Ok, I think I'll try to clear my morning tomorrow to give a deep look at the changes, thank you very much!

lucferbux commented:
@cam-garrison In the meantime, take a look at [this warnings](https://github.com/opendatahub-io/odh-dashboard/actions/runs/5397688926/jobs/9802633783?pr=1088) that you need to address before merging the PR. 
Runing locally `npm test` should tell you the linting warnings the CI is raising.

lucferbux commented:
/hold

lucferbux commented:
@cam-garrison I've done the first rundown, and things seems to be working fine, today is feature freeze and ill be out most of next week.
I'm gonna hold this until I come back from my PTO so we have plenty of time next sprint to test if there's any side effects in the integration.

cc @andrewballantyne

cam-garrison commented:
Hey @lucferbux have you gotten the chance to do any more testing with this PR? Please lmk if you have any questions on how to get things running!

lucferbux commented:
> Hey @lucferbux have you gotten the chance to do any more testing with this PR? Please lmk if you have any questions on how to get things running!

Yes @cam-garrison I'm sorry, I've been behind the schedule all week long, I'm aiming to review it today or tomorrow!

lucferbux commented on backend/src/routes/api/namespaces/namespaceUtils.ts:
```diff
@@ -60,13 +61,19 @@ export const applyNamespaceChange = async (
         'modelmesh-enabled': 'true',
       };
+      annotations = {
+        'opendatahub.io/service-mesh': String(!disableServiceMesh),
```
As @andrewballantyne mentioned before, this is problematic because it will get both `false` and `undefined`.

We need to implement something like `featureFlagEnabled` in the backend: https://github.com/opendatahub-io/odh-dashboard/blob/3843c8222bc330feb44d0e0cf08ae37747ae7467/frontend/src/utilities/utils.ts#L8-L9

cam-garrison replied:
I've added `featureFlagEnabled()` to the backend and used it in all places where we were previously pulling the value directly. Take a look here 93aa3ae

lucferbux commented on backend/src/routes/api/notebooks/utils.ts:
```diff
@@ -27,12 +29,24 @@ export const getNotebookStatus = async (
+        host = route.spec.host;
+      }
+    }
+    if (host) {
```
You can save the check above if you do something like:

```sugestion
if(route?.spec.host) {
    newNotebook = await patchNotebookRoute(fastify, route.spec.host, namespace, notebookName).catch((e) => {
```

cam-garrison replied:
I think my changes in latest commit 93aa3ae improve the logic there, but I don't thinkk using  

```
if(route?.spec.host) {
    newNotebook = await patchNotebookRoute(fastify, route.spec.host, namespace, notebookName).catch((e) => {
```

would work in this case since in the if (SM) statement we fetch the host directly, and else (no SM) we fetch the host from the route struct. Am I misinterpreting your comment?

cam-garrison replied:
Yes - marking as resolved, feel free to reopen if needed.

lucferbux reviewed (changes requested):
I need to do another rundown and test this with the latest changes in https://github.com/opendatahub-io/odh-manifests/pull/818

This first past was only for the code itself, I'll get to you again once I check latest behavior!

cam-garrison commented:
Thanks for the initial review @lucferbux  - working on implementing these changes now. For the squash - would it be sufficient to do a Squash and Merge? Not sure if there is some advantage to squashing beforehand that I don't know about.

lucferbux commented:
> Thanks for the initial review @lucferbux - working on implementing these changes now. For the squash - would it be sufficient to do a Squash and Merge? Not sure if there is some advantage to squashing beforehand that I don't know about.

@cam-garrison we use tide as our merging bot. In the past we had it configured to squash and merge, but things have changed recently (we might go back to the squash and merge when contributing into main) but as for now,  it would be super nice if you can squash everything to keep a clean history. 

Let me know if you have any question about that and thanks in advance!

cam-garrison commented:
@lucferbux Gotcha - sure that sounds fine :+1: I will plan on squashing all the changes after review is complete just to avoid doing so multiple times.

lucferbux commented:
Ok, so im gonna add the review asap

> @lucferbux Gotcha - sure that sounds fine 👍 I will plan on squashing all the changes after review is complete just to avoid doing so multiple times.

ok, I'm gonna try to have a review for today!

lucferbux commented:
@cam-garrison I didn't have the time today, I'll try to have it tomorrow, I'm sorry for the inconvenience.

lucferbux commented:
Hold until next release, we are focusing on just bug fixing for this sprint

bartoszmajsak commented:
> Hold until next release, we are focusing on just bug fixing for this sprint

@lucferbux I do understand the reasons for holding the merge back but I would very much appreciate it if someone from the team could review and share feedback about the latest code changes when they find the time.

lucferbux commented:
> > Hold until next release, we are focusing on just bug fixing for this sprint
> 
> @lucferbux I do understand the reasons for holding the merge back but I would very much appreciate it if someone from the team could review and share feedback about the latest code changes when they find the time.

Yes @bartoszmajsak I'm back after my PTO, I'm still catching up, this is a top priority for me this week, I'll try to have it for today or tomorrow

lucferbux commented:
I've added https://github.com/opendatahub-io/odh-dashboard/issues/1838 as a tracker, this should go to incubation first, changed targeted branch to `f/ossm`

lucferbux commented on backend/src/routes/api/notebooks/utils.ts:
```diff
@@ -27,12 +29,24 @@ export const getNotebookStatus = async (
+      host = route?.spec.host;
+    }
+    if (host) {
+      newNotebook = await patchNotebookRoute(fastify, host, namespace, notebookName).catch((e) => {
```
Just to add more context here, there are certain rules that can be applied to improve the code, some of them are just to improve the syntax others are to avoid side effects (such as mixing catches and awaits). Let me show you an example of this same function refactored to improve the sintax:

```javascript
export const getNotebookStatus = async (
  fastify: KubeFastifyInstance,
  namespace: string,
  name: string,
): Promise<{ notebook: Notebook; isRunning: boolean; podUID: string }> => {
  const notebook = await getNotebook(fastify, namespace, name);
  const hasStopAnnotation = !!notebook?.metadata.annotations?.['kubeflow-resource-stopped'];
  const [isRunning, podUID] = hasStopAnnotation
    ? [false, '']
    : await listNotebookStatus(fastify, namespace, name);

  const notebookName = notebook?.metadata.name;

  if (isRunning && !notebook?.metadata.annotations?.['opendatahub.io/link']) {
    const enableServiceMesh = featureFlagEnabled(
      getDashboardConfig().spec.dashboardConfig.disableServiceMesh,
    );
    const getHost = enableServiceMesh ? getServiceMeshGwHost : getRoute;
    const logError = (e: Error) =>
      fastify.log.warn(`Failed getting route ${notebookName}: ${e.message}`);

    try {
      const hostObj = await getHost(fastify, namespace, notebookName);
      if (typeof hostObj === 'object' && hostObj !== null && 'spec' in hostObj) {
        const host = hostObj.spec.host;
        if (host) {
          try {
            const newNotebook = await patchNotebookRoute(fastify, host, namespace, notebookName);
            return { notebook: newNotebook, isRunning, podUID };
          } catch (e) {
            fastify.log.warn(`Failed patching route to notebook ${notebookName}: ${e.message}`);
            return { notebook, isRunning, podUID };
          }
        }
      }
    } catch (e) {
      logError(e);
    }
  }
};
```

Take the logic with a grain of salt, it was just a quick refactor to display some techniques (avoiding using a lot of mutable variables, leveraging some logic to typescript's types and more)

lucferbux commented on backend/src/utils/notebookUtils.ts:
```diff
@@ -68,6 +69,35 @@ export const getRoute = async (
+    throw error;
+  });
+
+  const body = kubeResponse.body as unknown;
+  const typedResponse = body as V1Namespace;
+
+  const annotations = typedResponse.metadata?.annotations;
```
Why? `readNamespace` is already returning a typed response, you don't need to cast anything here, cast as little as possible, in this case you can have:


` const annotations = kubeResponse.body.metadata?.annotations;`

cam-garrison replied:
Good point, addressed here 

https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/ca51ef37fc25d64af1321ebfa40478a10b4fafc7

lucferbux commented on frontend/src/api/k8s/notebooks.ts:
```diff
@@ -287,9 +325,10 @@ export const updateNotebook = (
 ): Promise<NotebookKind> =>
   new Promise((resolve, reject) =>
-    createNotebook(data, username).then((notebook) =>
+    createNotebook(data, username, dashboardConfig).then((notebook) =>
```
Same with `createNotebook`

cam-garrison replied:
agree, addressed here https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/74b4c79c8c2e8b11c5e759bd158a7dc4462de224

lucferbux commented on frontend/src/api/k8s/routes.ts:
```diff
@@ -14,3 +14,14 @@ export const getRoute = (
+  const queryOptions = {
+    name: namespace,
+  };
+  const project = await k8sGetResource<NamespaceKind>({ model: NamespaceModel, queryOptions });
```
nit: In the frontend we tend to use Promises (.then()/.catch()) instead of async await, just as general rule.

lucferbux replied:
Also, if you want to get the project, don't use `NamespaceKind`, we already have `ProjectKind` in place

lucferbux commented on frontend/src/pages/projects/notebook/useRouteForNotebook.ts:
```diff
@@ -19,12 +21,17 @@ const useRouteForNotebook = (
-        getRoute(notebookName, projectName)
-          .then((route) => {
+        // if not using service mesh fetch openshift route, otherwise get Istio Ingress Gateway route
+        const getRoutePromise = dashboardConfig.spec.dashboardConfig.disableServiceMesh
```
use the utlity function `featureFlagEnabled` to avoid confusion here or either use a const as I mentioned above.

cam-garrison replied:
this file is outdated - not sure why it popped up - I have used the utility function [here](https://github.com/opendatahub-io/odh-dashboard/blob/c10666359453e5ac902b263c26ea6bb3c06659ca/frontend/src/pages/projects/notebook/useRouteForNotebook.ts#L25).

lucferbux replied:
great, thanks!

lucferbux reviewed (changes requested):
Hi Cam, I'm sorry for the late review, check these changes, I've managed to test a first pass, the logic is more in place now, I'll try to test edge cases in order to thoroughly test the logic here.

Also, I've changed the targeted branch and created a tracker, it should be safer to have this first in incubation, to avoid issues in rhods first.

cam-garrison commented on backend/src/utils/notebookUtils.ts:
```diff
@@ -444,7 +483,14 @@ export const createNotebook = async (
   }
 
-  notebookAssembled.metadata.annotations['notebooks.opendatahub.io/inject-oauth'] = 'true';
+  const enableServiceMesh = featureFlagEnabled(config.spec.dashboardConfig.disableServiceMesh);
```
true - https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/6b12f82bb750d4604ab41ab616b5ce133bd854f1

removed const and added the call here instead in 1 commit.

lucferbux replied:
great!

cam-garrison commented on backend/src/utils/notebookUtils.ts:
```diff
@@ -419,6 +456,8 @@ export const createNotebook = async (
   url: string,
   notebookData?: NotebookData,
 ): Promise<Notebook> => {
+  const config = getDashboardConfig();
```
Agree, no need for this - removed here https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/6b12f82bb750d4604ab41ab616b5ce133bd854f1

cam-garrison commented on frontend/src/api/k8s/notebooks.ts:
```diff
@@ -242,9 +278,10 @@ export const startNotebook = async (
   canEnablePipelines?: boolean,
 ): Promise<NotebookKind> => {
-  const notebook = assembleNotebook(data, username, canEnablePipelines);
+  const notebook = assembleNotebook(data, username, dashboardConfig, canEnablePipelines);
```
agree, addressed here https://github.com/opendatahub-io/odh-dashboard/pull/1088/commits/74b4c79c8c2e8b11c5e759bd158a7dc4462de224

cam-garrison commented on frontend/src/api/k8s/notebooks.ts:
```diff
@@ -210,10 +237,19 @@ export const startNotebook = async (
 ): Promise<NotebookKind> => {
-  const patches: Patch[] = [];
-  patches.push(startPatch);
+  const enableServiceMesh = featureFlagEnabled(
```
I am not sure I understand. Are you suggesting creating a util function that getsAppContext, then calls featureFlagEnabled to parse the service mesh flag (therefore taking no inputs and returning this bool)?

lucferbux replied:
Yes, I'm talking about something like this: https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/pages/projects/projectSharing/utils.ts#L7 is not mandatory but it might be nice to have

cam-garrison commented:
Thx for the review! I think I've addressed all points now - left one followup question where I wasn't sure. 

Re: 

> I've added https://github.com/opendatahub-io/odh-dashboard/issues/1838 as a tracker, this should go to incubation first, changed targeted branch to f/ossm

What does "moving to incubation" mean exactly? testing by QE? Will there be some publicly available build available? Just want to know what the incubation phase entails.

Thanks!

cam-garrison commented:
@lucferbux can we re-target this to incubation [branch](https://github.com/opendatahub-io/odh-dashboard/tree/incubation)?

cam-garrison commented:
Added manifests changes from #1900 into this PR:
See closed PR desc for full reasoning, TLDR: since v2 operator pulls manifests from dashboard, need to add requisite service mesh overlay.

andrewballantyne commented:
As I noted in Slack: https://github.com/opendatahub-io/odh-dashboard/blob/main/docs/process-definition/branches.md our branching strategy uses feature branches first. Going directly to incubation removes the layer of indirection that allows it to ever merge back to `main` & onto downstream products like RHODS.

lucferbux commented on manifests/overlays/service-mesh/OWNERS:
```diff
@@ -0,0 +1,10 @@
+# Each list is sorted alphabetically, additions should maintain that order
```
why do we have this OWNERS files in the manifest, I don't understand

andrewballantyne replied:
OWNERS files can be placed at lower levels to support "just these files" changing. So if something under the `overlays/service-mesh` folder is the only thing to change, they would need need the repo ownership to step in to review it. It's additive on 

I think this is fine for this use-case... We should definitely avoid OWNERS files over code so that we can still approve the holistic nature of that.

bartoszmajsak replied:
> why do we have this OWNERS files in the manifest, I don't understand

Our line of thinking is that we offer help and knowledge to keep this thing up-to-date and evolving. So if anything will be proposed to change here we will be pinged

lucferbux commented on manifests/overlays/minimal/README.md:
```diff
@@ -0,0 +1,4 @@
+## Dashboard Minimal overlay
```
I don't see the harm on having a minimal overlay, but maybe we would want to have this in other PR?

cc @andrewballantyne @christianvogt what do you think? Not sure if we should stick this, which is related to a deployment of our repo with this same PR.

andrewballantyne replied:
This is a neat idea... I'm not against new overlays like this. @cam-garrison was this something you used in your testing/coding?

bartoszmajsak replied:
Yeah, this can be removed. It is anyway not possible to combine overlays like it was with `KfDef` magic, so better to make it separate and shrink this PR.

cam-garrison replied:
Yes, we used this for testing/running in CRC specifically. +1 on bartosz's point though, can be removed especially since in v2 operator it is not as easy to apply multiple overlays.

lucferbux commented on manifests/base/_mesh_overlay-delete-oauth-config-patch.yaml:
```diff
@@ -0,0 +1,6 @@
+# DO NOT DELETE. Check kustomization.yaml comments in overlays/service-mesh
+$patch: delete
```
mmm not 100% sure, but if we delete the `dashboard-oauth-config`, should we do something with the volume `dashboard-oauth-config-generated`?

cam-garrison replied:
yes, we handle this in the patches to the deployment https://github.com/opendatahub-io/odh-dashboard/blob/2d2244cfb67a566504530a66b767aa540de9aca4/manifests/overlays/service-mesh/patches/deployment-patch.yaml#L5

andrewballantyne replied:
Would we not want this in your overlay? This is in the base? I'll admit I'm kinda a noob when it comes to the power of the kustomization, but it seems odd to have this in the base for everything.

I'm confused why there are two of these -- one here & one in the overlay 🤔

cam-garrison replied:
Unfortunately this is due to the weakness/limitation/bug of kustomize - not the power. From comment in overlays/service-mesh/kustomization.yaml
```
# when using bases:
# - ../../base
# the entire directory is truncated and the lookup is performed in base/
# for this reason the files must be there as well. with the same content, as they will be in fact used
```

cam-garrison replied:
This comment is not accurate for v2 operator - we can safely remove this. Will do this during/after our call :+1:

lucferbux reviewed (changes requested):
Just small nits, I'm not 100% sure about all the config files, i lack a little bit more context for that, and I haven't tested that yet, but so far code looks great.

Please squash before we can merge, thanks a lot for all the work @cam-garrison

andrewballantyne commented on frontend/src/api/k8s/notebooks.ts:
```diff
@@ -210,10 +233,19 @@ export const startNotebook = async (
+
+  const patches: Patch[] = [
+    startPatch,
+    getInjectOAuthPatch(enableServiceMesh),
+    getServiceMeshPatch(enableServiceMesh),
+    getProxyInjectPatch(enableServiceMesh),
```
These seem all very related, we may want to look at cleaning that up in the future to make it one "set of patches" rather than 3 separate ones. Just for code cleanness.

andrewballantyne commented on frontend/src/pages/projects/notebook/useRouteForNotebook.ts:
```diff
@@ -19,12 +22,20 @@ const useRouteForNotebook = (
+          dashboardConfig.spec.dashboardConfig.disableServiceMesh,
+        );
+        // if not using service mesh fetch openshift route, otherwise get Istio Ingress Gateway route
+        const getRoutePromise = !enableServiceMesh
+          ? getRoute(notebookName, projectName).then((route) => route?.spec.host)
+          : getServiceMeshGwHost(projectName);
```
Interesting idea... using the `getRoute` when `getServiceMeshGwHost` returns an invalid value?

andrewballantyne commented on frontend/src/pages/projects/notebook/useRouteForNotebook.ts:
```diff
@@ -46,7 +57,7 @@ const useRouteForNotebook = (
       clearTimeout(watchHandle);
     };
-  }, [notebookName, projectName, isRunning]);
+  }, [notebookName, projectName, isRunning, dashboardConfig]);
```
Dashboard value may churn due to being a K8s object... the flag for enableServiceMesh is a boolean and thus is a primitive and even if the Dashboard changes the value is unlikely to change so easily.

`obj === obj2` -- even if they have the same values, the two are not equal unless they have the same memory reference. 
`boolA === boolB` -- always will compare value because they are primitives.

andrewballantyne commented on frontend/src/utilities/notebookControllerUtils.ts:
```diff
@@ -213,7 +213,7 @@ export const useNotebookRedirectLink = (): (() => Promise<string>) => {
         getRoute(notebookNamespace, routeName)
           .then((route) => {
-            resolve(`https://${route.spec.host}/notebook/${notebookNamespace}/${routeName}`);
+            resolve(`https://${route.spec.host}/notebook/${notebookNamespace}/${routeName}/`);
```
this is curious -- why did this change happen? I assumed the route efforts was un-impacted by this new host information

christianvogt replied:
I saw the same thing but when i searched our code i found that other areas that create such a URL append `/`.

cam-garrison replied:
The change here specifically is because of the way that the kf notebook-controller generates istio `VirtualService` resources  (function [here](https://github.com/opendatahub-io/kubeflow/blob/e5d179d9e8ecfea666be4fe0dc2f0d55505673e9/components/notebook-controller/controllers/notebook_controller.go#L516)) - essentially since the virtual service has the trailing `/` we have to add it here as well.

andrewballantyne replied:
Cameron and I talked and this is not needed for the ServiceMesh flow. We agreed this is not needed.

andrewballantyne commented on manifests/base/_mesh_overlay_delete-oauth-client-patch.yaml:
Same question here -- I understand these come back with the Operator, but it seems a little destructive when you're not using the OSSM path 🤔

andrewballantyne commented on manifests/overlays/service-mesh/patches/oauth-client-patch.yaml:
What's the added value of these files? Seems this is the 3rd iteration of the _same_(?) operation?

cam-garrison replied:
Thanks for pointing this out - this and `overlays/service-mesh/patches/oauth-config-patch.yaml` can be safely removed. I believe these may be remnants from an old rebase.

andrewballantyne commented:
For the record -- I asked Cam to merge it with a squash to save the time working around the merge commits from main. 

He had powers because of the [org-management permissions](https://github.com/opendatahub-io/org-management/blob/7f732bbfe414daeb4fb03a49fa0bd5f461a2081b/config/opendatahub-io/org.yaml#L528C15-L528C15) -- I'll look to be removing this shortly.
